### PR TITLE
Implement sort in PHP

### LINKED ISSUE #2: Implement sort(1) from scratch in PHP
... so that `php sort.php < in.txt` does the same as `sort < in.txt` (implement the sort algo yourself!)

### PR DESCRIPTION
**Pull request by yhabteab**
fixes #2

### CODE REVIEW

File: sort.php [hunk 1]
```diff
@@ -0,0 +1,26 @@
+<?php
+
+ // File which we want to sort
+$array = file('in.txt');
```
Comment (Al2Klimov) [inline sort.php]: You don't need this variable – you can just do:

```php
<?php
a(b());
```

... instead of:

```php
<?php
$c = b();
a($c);
```

File: sort.php [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+function bubbleSort($array)
```
Comment (Al2Klimov) [inline sort.php]: Don't name your variables after PHP keywords, i.e. "array".

File: sort.php [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+            $yonas++;
+            $tmp = $array[$i]; // $temp = $array[0] value
+            if ($array[$i] < $array[$j]) {
+                $array[$i] = $array[$j]; // $array[0] = $array[1] value
+                $array[$j] = $tmp; // $array[1] = $temp value
+            }
+        }
+    }
+
+    return implode('', $array);
```
Comment (Al2Klimov) [inline sort.php]: Please return an array and implode it outside of this function.

File: sort.php [hunk 4]
```diff
@@ -0,0 +1,26 @@
+<?php
+
+ // File which we want to sort
+$array = file('in.txt');
+
+function bubbleSort($array)
+{
+    $length = count($array);
+    $comparison = 0;
+    $yonas = 0;
```
Comment (Al2Klimov) [inline sort.php]: What's this variable for?

File: sort.php [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+    for ($i = 1; $i < $length; $i++) {
+        $comparison++;
+        for ($j = 0; $j < $length; $j++) {
+            $yonas++;
+            $tmp = $array[$i]; // $temp = $array[0] value
+            if ($array[$i] < $array[$j]) {
+                $array[$i] = $array[$j]; // $array[0] = $array[1] value
+                $array[$j] = $tmp; // $array[1] = $temp value
```
Comment (Al2Klimov) [inline sort.php]: IMAO these comments are too much.

**Review by Al2Klimov — CHANGES_REQUESTED** [review]

File: sort.php (hunk 1 above)
Comment (Al2Klimov) [reply sort.php]: Still not fixed.

File: sort.php [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+                $sort_array[$i] = $sort_array[$j];
+                $sort_array[$j] = $tmp;
+            }
+        }
+    }
+
+    return $sort_array;
+}
+
+$sorted = implode("\n", bubbleSort(stream_get_contents(STDIN)));
```
Comment (Al2Klimov) [inline sort.php]: Why did you change this to `stream_get_contents(STDIN)`?

File: sort.php [hunk 7]
```diff
@@ -0,0 +1,23 @@
+<?php
+
+function bubbleSort($sort_array)
+{
+    $sort_array = explode("\n", $sort_array);
+    $sort_arrayIndex = count($sort_array);
+
+    for ($i = 0; $i < $sort_arrayIndex; $i++) {
+
```
Comment (Al2Klimov) [inline sort.php:15]: This line is a bit too much.

File: sort.php [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+        for ($j = 0; $j < $sort_arrayIndex ; $j++) {
+            if ($sort_array[$i] < $sort_array[$j]) {
+                $tmp            = $sort_array[$i];
+                $sort_array[$i] = $sort_array[$j];
+                $sort_array[$j] = $tmp;
```
Comment (Al2Klimov) [inline sort.php:26]: The `=`s of the assignments have not to be aligned in a column like this.

**Review by Al2Klimov — CHANGES_REQUESTED** [review]

File: sort.php [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+            }
+        }
+    }
+
+    return $sort_array;
+}
+
+echo implode("", bubbleSort(file('php://stdin')));
+
```
Comment (Al2Klimov) [inline sort.php]: Again empty one line too many.

**Review by Al2Klimov — CHANGES_REQUESTED** [review]

File: sort.php [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+                $tmp = $sort_array[$i];
+                $sort_array[$i] = $sort_array[$j];
+                $sort_array[$j] = $tmp;
+            }
+        }
+    }
+
+    return $sort_array;
+}
+echo implode("", bubbleSort(file('php://stdin')));
```
Comment (Al2Klimov) [inline sort.php]: Unfortunately now you have one line too less. The "echo implode..." shall be the last line, but there shall not be the "No newline at and of file" icon. I.e. every line shall be \\n-terminated.

File: sort.php [hunk 11]
```diff
@@ -0,0 +1,22 @@
+<?php
+
+function bubbleSort($sort_array)
+{
+    $array_indexes = count($sort_array);
+    $lastIndex =  $array_indexes - 1;
+    if (strlen($sort_array[$lastIndex]) < 2) {
```
Comment (Al2Klimov) [inline sort.php]: For the sake of readability please separate these two lines ("$lastIndex..." and "if...") by an empty line.

File: sort.php [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+        $sort_array[$lastIndex] = $sort_array[$lastIndex] . "\n";
+    }
+    for ($i = 0; $i < $array_indexes; $i++) {
```
Comment (Al2Klimov) [inline sort.php:21]: Please also separate these ones.

File: sort.php (hunk 10 above)
Comment (Al2Klimov) [inline sort.php]: Please also separate these ones.

**Review by Al2Klimov — CHANGES_REQUESTED** [review]

**Review by Al2Klimov — CHANGES_REQUESTED** [review]
`php sort.php < /dev/null` output differs from `sort < /dev/null`. Please fix that.

**Review by Al2Klimov — CHANGES_REQUESTED** [review]
The outputs of

```bash
sort <<EOF
abc
def
EOF
```

and

```bash
php sort.php <<EOF
abc
def
EOF
```

differ. Please fix that.

**Review by Al2Klimov — CHANGES_REQUESTED** [review]
The outputs of

```bash
sort <<EOF
abc

EOF
```

and

```bash
php sort.php <<EOF
abc

EOF
```

differ. Please fix that.

File: sort.php [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+    if (empty($sort_array)) {
+        return [];
+    }
+
+    $array_indexes = count($sort_array);
+
+    for ($i = 0; $i < $array_indexes; $i++) {
+        if (in_array("", $sort_array)) {
+            continue;
+        }
```
Comment (Al2Klimov) [inline sort.php]: Please explain what exactly this if does.

Comment (yhabteab) [reply sort.php]: Es wird nach jeder Schleifen Iteration überprüft, ob ein leeres String in $sort_array gibt. Wenn es true ist, dann ignoriert er den leeren String und bricht die Iteration und fängt von vorne an.

Comment (Al2Klimov) [reply sort.php]: Nein, nicht _nach_, sondern _zu Beginn_. In welchem Fall hast Du dort einen leeren String?

**Comment by yhabteab** [comment]
Is the function now fully implemented? You haven't answered me since before yesterday, so I assume that you haven't found a bug in the function.

**Comment by Al2Klimov** [comment]
IMO your last commit, 7291c00ca4eca5b33149a3e4adbc505c15bb9916 , was wrong because it reverted https://github.com/Al2Klimov/trainee-php-sort/pull/14/commits/cdbd33cdf0d2a00b382663f257a11b303b50ac7a#diff-ee5ea956c1778d872e0fc7251f5384a3R7-R9 . That appending of a \n was right, but the condition was quite wrong. Please let me know once you come back to this project (no other tasks to do) and I'll explain the details to you.

File: sort.php [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+function bubbleSort($sort_array)
+{
+    $array_indexes = count($sort_array);
+
+    for ($i = 0; $i < $array_indexes; $i++) {
+        if (strpos($sort_array[$i], "\n") === false) {
+            $sort_array[$i] .= "\n";
+        }
+
+        for ($j = 0; $j < $array_indexes; $j++) {
```
Comment (Al2Klimov) [inline sort.php]: Are you sure $j's start and end values are correct?

**Review by Al2Klimov — CHANGES_REQUESTED** [review]

File: sort.php (hunk 14 above)
Comment (Al2Klimov) [reply sort.php]: I.e. 0 and $array_indexes.

Comment (yhabteab) [reply sort.php]: yes absolutely, if you want me to use a different variant i can change it but the way it is works for me quite well

File: sort.php [hunk 15]
```diff
@@ -0,0 +1,35 @@
+<?php
+
+function bubbleSort($sort_array)
+{
+    if (empty($sort_array)) {
+        return [];
+    }
```
Comment (Al2Klimov) [inline sort.php:11]: What's this for?

File: sort.php [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+            }
+        }
+
+        if (! $sorted) {
+            break;
+        }
+
+        if (strpos($sort_array[$i], "\n") === false) {
+            $sort_array[$i] .= "\n";
+        }
```
Comment (Al2Klimov) [inline sort.php]: Please put this out of the bubbleSort function (and do it before sorting) as this has nothing to do with bubble sort.

**Review by Al2Klimov — CHANGES_REQUESTED** [review]

File: sort.php (hunk 15 above)
Comment (yhabteab) [reply sort.php:11]: If the array should be empty then it will return immediately without doing anything else. We don't need to check anything if it's empty.

Comment (Al2Klimov) [reply sort.php:11]: Even without this, the function returns more-or-less immediately. Please explain why.

Comment (yhabteab) [reply sort.php:11]: Because after each loop we check if something was swapped, if not then we abort and return what is in the array no matter if they are already sorted or the array is completely empty.

Comment (Al2Klimov) [reply sort.php:11]: So please remove the if.

File: sort.php (hunk 16 above)
Comment (yhabteab) [reply sort.php]: Where should I implement this function before or after bubblesort function?

Comment (Al2Klimov) [reply sort.php]: As I said: before.

File: sort.php [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+
+    return $arr;
+}
+
+function bubbleSort($sort_array)
+{
+    $array_indexes = count($sort_array);
+    $sorted = false;
+
+    $checked = checkForNextLine($sort_array);
```
Comment (Al2Klimov) [inline sort.php]: Please also call it in the "main function", not in the bubbleSort function as it has nothing to do with bubble sort.

**Review by Al2Klimov — CHANGES_REQUESTED** [review]

File: sort.php (hunk 17 above)
Comment (yhabteab) [reply sort.php]: Which main function? we don't have a main function.

Comment (Al2Klimov) [reply sort.php]: I mean the same place where you call the bubbleSort function.

**Review by Al2Klimov — CHANGES_REQUESTED** [review]
Write a comment in this PR (not in the code) with a so-called "test protocol" – compare the outputs of the standard sort and yours.

**Comment by yhabteab** [comment]
###Test protocol:

In this test we will compare our self-implemented sort func with the standard sort func which looks like this...

Here we see the output of both functions:

If you run this commands on your command line the output is exact the same.
php sort.php < in.txt          |        sort < in.txt
------------              |           -------------
 0 | 0
 1  | 1
 2  | 2
 3  | 3
 4  | 4
 5  | 5
 6  | 6
 7  | 7
 8  | 8
 9  | 9  

or if you run this command..
```
php sort.php <<EOF                                
abc
EOF
```
and this command
```
sort <<EOF
abc
EOF
```
the output schould be exact the same.
Last thing I want you to do is run this on your command line...
```
php sort.php < /dev/null
```
```
sort < /dev/null
```
 nothing should come out.

**Review by Al2Klimov — CHANGES_REQUESTED** [review]
* [ ] `git reset --soft HEAD~$(git rev-list --count $(git merge-base HEAD master)..HEAD^)`
* [ ] `git commit --amend --no-edit`
* [ ] `git push -f origin feature/sort-2`

**Review by Al2Klimov — APPROVED** [review]
